src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+#pragma warning disable 420
+
+namespace DotNetty.Common.Utilities
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Diagnostics.Contracts;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using DotNetty.Common.Concurrency;
+    using DotNetty.Common.Internal;
+    using DotNetty.Common.Internal.Logging;
+
+    public sealed class HashedWheelTimer : ITimer
+    {
+        static readonly IInternalLogger Logger =
+            InternalLoggerFactory.GetInstance<HashedWheelTimer>();
+
+        static int instanceCounter;
+        static int warnedTooManyInstances;
+
+        const int InstanceCountLimit = 64;
+
+        readonly Worker worker;
+        readonly XThread workerThread;
+        readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+        const int WorkerStateInit = 0;
+        const int WorkerStateStarted = 1;
+        const int WorkerStateShutdown = 2;
+        int workerStateVolatile = WorkerStateInit; // 0 - init, 1 - started, 2 - shut down
+
+        readonly long tickDuration;
+        readonly HashedWheelBucket[] wheel;
+        readonly int mask;
+        readonly CountdownEvent startTimeInitialized = new CountdownEvent(1);
+        readonly IQueue<HashedWheelTimeout> timeouts = PlatformDependent.NewMpscQueue<HashedWheelTimeout>();
+        readonly IQueue<HashedWheelTimeout> cancelledTimeouts = PlatformDependent.NewMpscQueue<HashedWheelTimeout>();
+        internal long PendingTimeouts;
+        readonly long maxPendingTimeouts;
+        long startTimeVolatile;
+
+        public HashedWheelTimer()
+            : this(TimeSpan.FromMilliseconds(100), 512, -1)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new timer.
+        /// </summary>
+        /// <param name="tickDuration">the duration between tick</param>
+        /// <param name="ticksPerWheel">the size of the wheel</param>
+        /// <param name="maxPendingTimeouts">The maximum number of pending timeouts after which call to
+        /// <c>newTimeout</c> will result in <see cref="RejectedExecutionException"/> being thrown.
+        /// No maximum pending timeouts limit is assumed if this value is 0 or negative.</param>
+        /// <exception cref="ArgumentException">if either of <c>tickDuration</c> and <c>ticksPerWheel</c> is &lt;= 0</exception>
+        public HashedWheelTimer(
+            TimeSpan tickDuration,
+            int ticksPerWheel,
```

Comment:
that would be inconvenient in calculations (due to need for conversion)

---
In reply to: [114473422](https://github.com/Azure/DotNetty/pull/242#discussion_r114473422) [](ancestors = 114473422)